tests/rados/stretch_cluster.py

```diff
@@ -0,0 +1,472 @@
+        return 1
+
+    # Setting the election strategy to connectivity mode
+    cmd = "/bin/ceph mon set election_strategy connectivity"
+    cephadm.shell([cmd])
+
+    # Sleeping for 5 sec for the strategy to be active
+    time.sleep(5)
+    init_mon_state = get_mon_details(node=cephadm)
+
+    # Checking if mon elections happened after changing election strategy
+    if mon_state["epoch"] > init_mon_state["epoch"]:
+        log.error("Election epoch not bumped up after setting the connectivity mode.")
+        return 1
+
+    # Checking updated election strategy in mon map
+    if init_mon_state["election_strategy"] != 3:
+        log.error(
+            f"Election strategy is not connectivity mode.\n Currently set {mon_state['election_strategy']}"
+        )
+        return 1
+    log.info("Enabled connectivity mode on the cluster")
+
+    log.info(f"selecting mon : {tiebreaker_node} as tie breaker monitor on site 3")
+    if not set_mon_sites(node=cephadm, tiebreaker_node=tiebreaker_node):
+        log.error("Failed to ad monitors into respective sites")
+        return 1
+
+    # All the existing pools should be automatically changed with stretch rule. Creating a test pool
+    pool_name = "test_pool_1"
+    if not create_pool(
+        node=cephadm, disable_pg_autoscale=True, pool_name=pool_name, pg_num=16
+    ):
+        log.error("Failed to create the replicated Pool")
+        return 1
+
+    log.info("Monitors added to respective sites. enabling stretch rule")
+    cmd = f"/bin/ceph mon enable_stretch_mode {tiebreaker_node.hostname} {stretch_rule_name} datacenter"
+    try:
+        cephadm.shell([cmd])
+    except Exception as err:
+        log.error(
+            f"Error while enabling stretch rule on the datacenter. Command : {cmd}"
+        )
+        log.error(err)
+        return 1
+
+    if get_mon_details(node=cephadm)["epoch"] < init_mon_state["epoch"]:
+        log.error("Election epoch not bumped up after Enabling strech mode")
+        return 1
+
+    if config.get("perform_add_capacity"):
+        pool_name = "test_stretch_pool"
+        if not create_pool(
+            node=cephadm,
+            disable_pg_autoscale=True,
+            pool_name=pool_name,
+            crush_rule=stretch_rule_name,
+        ):
+            log.error("Failed to create the replicated Pool")
+            return 1
+        do_rados_put(mon=client_node, pool=pool_name, nobj=100)
+
+        log.info("Performing add Capacity after the deployment of stretch cluster")
+        site_a_osds = [osd for osd in sorted_osds[0] if osd not in site_a_osds]
+        site_b_osds = [osd for osd in sorted_osds[1] if osd not in site_b_osds]
+
+        if not set_osd_sites(
+            node=cephadm,
+            site_a_osds=site_a_osds,
+            site_b_osds=site_b_osds,
+            all_osd_details=osd_details,
+        ):
+            log.error("Failed to move the OSD's into two sites")
+            return 1
+
+        time.sleep(10)
```

Comment:
It's not a good practice to blindly sleep. If we are waiting for an event to complete, can we check for that event and the progress to the next step.

If the event status cannot be retrieved then let's add a comment

Reply:
added comment